src/data/ast.rs

```diff
@@ -436,78 +436,64 @@ impl Declarator {
-                } else {
-                    format!("*{}{}", qs, name)
-                };
-                match **to {
-                    Array { .. } | Function { .. } => write!(f, "({})", pointer),
-                    // this is the only case when `*const (*p)()` can occur
-                    // TODO: maybe put a space after the qualifier if it's there?
-                    // clang-format disagrees with me about the space though
-                    End => write!(f, "{}", pointer),
-                    Pointer { .. } => write!(f, "{}", pointer),
+        for declarator_type in unrolled_type[..unrolled_type.len() - 1].iter().rev() {
+            match declarator_type {
+                DeclaratorType::Pointer { qualifiers, .. } => {
+                    write!(
+                        f,
+                        "(*{}",
+                        qualifiers
+                            .iter()
+                            .map(|q| format!("{} ", q))
+                            .collect::<Vec<_>>()
+                            .concat()
+                    )?;
                 }
-            }
-            Array { of, .. } => of.print_mid(name, f),
-            End | Function { .. } => {
-                if let Some(name) = name {
-                    write!(f, "{}", name)?;
-                }
-                Ok(())
+                DeclaratorType::Array { .. } | DeclaratorType::Function(_) => {}
+                DeclaratorType::End => unreachable!(),
             }
         }
-    }
-    fn print_post(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        use DeclaratorType::*;
-        match self {
-            Pointer { to, .. } => to.print_post(f),
-            Array { of, size } => {
-                write!(f, "[")?;
-                if let Some(expr) = size {
-                    write!(f, "{}", expr)?;
+        if let Some(name) = name {
+            write!(f, "{}", name)?;
+        }
+        for declarator_type in unrolled_type[..unrolled_type.len() - 1].iter() {
+            match declarator_type {
+                DeclaratorType::Array { size, .. } => {
+                    if let Some(size) = size {
+                        write!(f, "[{}]", size)?;
+                    } else {
+                        write!(f, "[]")?;
+                    }
                 }
-                write!(f, "]")?;
-                of.print_post(f)
-            }
-            Function(FunctionDeclarator {
-                params, varargs, ..
-            }) => {
-                write!(f, "({}", joined(params, ", "))?;
-                if *varargs {
-                    write!(f, ", ...")?;
+                DeclaratorType::Function(function_definition) => {
+                    write!(f, "({}", joined(function_definition.params.iter(), ", "))?;
+                    if function_definition.varargs {
+                        write!(f, ", ...")?;
+                    }
+                    write!(f, ")")?;
+                }
```

Comment:
This is a declarator, not a definition. A declarator is `f(int)`, a definition is `int f(int a) { return 1; }`. All definitions are also declarations, but not all declarations are definitions, e.g. `int f(int);`.